src/platform_impl/web/web_sys/event.rs

```diff
@@ -61,192 +62,83 @@ pub fn mouse_scroll_delta(event: &WheelEvent) -> Option<MouseScrollDelta> {
-        "Unlabeled" => VirtualKeyCode::Unlabeled,
-        "AudioVolumeDown" => VirtualKeyCode::VolumeDown,
-        "AudioVolumeUp" => VirtualKeyCode::VolumeUp,
-        "Wake" => VirtualKeyCode::Wake,
-        "WebBack" => VirtualKeyCode::WebBack,
-        "WebFavorites" => VirtualKeyCode::WebFavorites,
-        "WebForward" => VirtualKeyCode::WebForward,
-        "WebHome" => VirtualKeyCode::WebHome,
-        "WebRefresh" => VirtualKeyCode::WebRefresh,
-        "WebSearch" => VirtualKeyCode::WebSearch,
-        "WebStop" => VirtualKeyCode::WebStop,
-        "Yen" => VirtualKeyCode::Yen,
-        _ => return None,
-    })
+fn cached_string<S: AsRef<str>>(string: S) -> &'static str {
+    use std::collections::HashSet;
+    use std::sync::Mutex;
+
+    use once_cell::sync::Lazy;
+
+    static STRING_CACHE: Lazy<Mutex<HashSet<&'static str>>> =
+        Lazy::new(|| Mutex::new(HashSet::new()));
+
+    let string = string.as_ref();
+    let mut cache = STRING_CACHE.lock().unwrap();
+    if let Some(string) = cache.get(string) {
+        string
+    } else {
+        // borrowck couldn't quite figure out this one on its own
+        let string: &'static str = Box::leak(String::from(string).into_boxed_str());
+        cache.insert(string);
+        string
+    }
 }
 
-pub fn keyboard_modifiers(event: &KeyboardEvent) -> ModifiersState {
-    let mut m = ModifiersState::empty();
-    m.set(ModifiersState::SHIFT, event.shift_key());
-    m.set(ModifiersState::CTRL, event.ctrl_key());
-    m.set(ModifiersState::ALT, event.alt_key());
-    m.set(ModifiersState::LOGO, event.meta_key());
-    m
+pub fn key_location(event: &KeyboardEvent) -> KeyLocation {
+    let location = event.location();
+    // As defined in the UIEvents specification
+    // https://w3c.github.io/uievents/#idl-keyboardevent
+    match location {
+        0 => KeyLocation::Standard,
+        1 => KeyLocation::Left,
+        2 => KeyLocation::Right,
+        3 => KeyLocation::Numpad,
+        _ => KeyLocation::Standard,
+    }
 }
 
-pub fn codepoint(event: &KeyboardEvent) -> char {
-    // `event.key()` always returns a non-empty `String`. Therefore, this should
-    // never panic.
-    // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key
-    event.key().chars().next().unwrap()
+// TODO: What should be done about `KeyboardEvent.isComposing`?
+
+pub fn keyboard_modifiers(key: &Key<'_>) -> ModifiersState {
+    match key {
+        Key::Shift => ModifiersState::SHIFT,
+        Key::Control => ModifiersState::CONTROL,
+        Key::Alt => ModifiersState::ALT,
+        Key::Super => ModifiersState::SUPER,
+        _ => ModifiersState::empty(),
+    }
 }
 
+// pub fn codepoint(event: &KeyboardEvent) -> char {
+//     // `event.key()` always returns a non-empty `String`. Therefore, this should
+//     // never panic.
+//     // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key
+//     event.key().chars().next().unwrap()
+// }
```

Comment:
As far as I understand this was used for the `keypress` event, which we don't use anymore.